lib/spack/spack/database.py

```diff
@@ -236,26 +247,40 @@ def prefix_lock(self, spec):
         if prefix not in self._prefix_locks:
             self._prefix_locks[prefix] = Lock(
                 self.prefix_lock_path,
-                spec.dag_hash_bit_prefix(bit_length(sys.maxsize)), 1)
+                start=spec.dag_hash_bit_prefix(bit_length(sys.maxsize)),
+                length=1,
+                default_timeout=self.package_lock_timeout)
 
         return self._prefix_locks[prefix]
 
     @contextlib.contextmanager
     def prefix_read_lock(self, spec):
         prefix_lock = self.prefix_lock(spec)
+        prefix_lock.acquire_read()
+
         try:
-            prefix_lock.acquire_read(60)
             yield self
-        finally:
+        except LockError:
+            raise
+        except (Exception, KeyboardInterrupt):
+            prefix_lock.release_read()
+            raise
+        else:
             prefix_lock.release_read()
 
     @contextlib.contextmanager
     def prefix_write_lock(self, spec):
         prefix_lock = self.prefix_lock(spec)
+        prefix_lock.acquire_write()
+
         try:
-            prefix_lock.acquire_write(60)
             yield self
-        finally:
+        except LockError:
```

Comment:
Do we need the check for `LockError` here since the `acquire_write` has now been hoisted out of the `try` block?  Is this specifically addressing the nested case?  It seems like this try/except/else block and the one below could use some comments, so that we at least don't forget what is going on here.

Reply:
@tgamblin there is now an explanatory comment: I didn't see a reason to do it other than the one you mentioned here, so I added a comment mentioning that specifically.